### PR TITLE
Start all applications via the shoehorn boot script

### PR DESCRIPTION
**Pull request by fhunleth**
This changes how Shoehorn starts OTP applications. Before this change,
the boot script would start Shoehorn and everything Shoehorn needs.
Shoehorn would then run `Application.ensure_all_started/1` on the
`:init` list and then on the main application.

The new way is to construct the boot script so that it starts
everything.

This has the following advantages:

1. The OTP Application start order is deterministic. If you want to see
   it, review `shoehorn.script`. `Application.ensure_all_started/1`
   might have been deterministic, but it wasn't possible to review
   offline.
2. It's faster. Loading applications via the boot script doesn't need to
   go through the Erlang module search path like
   `Application.ensure_all_started/1` does. Since the Erlang search path
   can be really long, this saves a lot of time on platforms that are slow
   at these searches. SquashFS is surprisingly slow.
3. It's possible to force an order on application loads that do not have
   any interdependencies. This provides consistency in load order when
   you add or remove dependencies. Shoehorn orders these alphabetically.
4. The main application no longer needs to be in the `:shoehorn`
   configuration. The `:app` key is ignored now.

As a bonus, this commit also lets you specify applications that should
be loaded last (`:last` option). This lets you defer applications to the
end that might delay user-perceived boot time. For example, starting
`:iex` is probably less important than starting networking and your
main application dependencies.

This commit also lets you specify additional dependencies between
applications.

Both of these new options are experimental, but are trivial to add now.
The `:init` list should work transparently for starting apps listed
there first.

### CODE REVIEW

**Comment by fhunleth** [comment]
@LostKobrakai I switched the implementation to use `:digraph`, added all of the dependencies as edges, and then added additional edges for the `:init` apps, and forced an alphabetical order on anything where the order was arbitrary.

File: lib/shoehorn/release.ex [hunk 1]
```diff
@@ -15,26 +15,170 @@ defmodule Shoehorn.Release do
...
+
+    dep_graph
+  end
+
+  # defp inspect_graph(dep_graph) do
+  #   Enum.each(:digraph.edges(dep_graph), fn e ->
+  #     {_, v1, v2, label} = :digraph.edge(dep_graph, e)
+  #     IO.puts("#{v1} -> #{v2} (#{label})")
+  #   end)
+  # end
```
Comment (jjcarstens) [inline lib/shoehorn/release.ex:149]: Is this needed?

File: lib/shoehorn/release.ex [hunk 2]
```diff
@@ -15,26 +15,170 @@ defmodule Shoehorn.Release do
...
+
+  defp order_dependencies(_, []), do: :ok
+
+  defp order_dependencies(dep_graph, [dep | rest]) do
+    Enum.each(rest, &:digraph.add_edge(dep_graph, dep, &1, :alpha))
+    order_dependencies(dep_graph, rest)
+  end
```
Comment (jjcarstens) [inline lib/shoehorn/release.ex:156]: Does `digraph` keep some sort of state? This seems like the only function that looks to adjust the `dep_graph`, but doesn't return an altered version of it

Comment (fhunleth) [reply lib/shoehorn/release.ex:156]: `:digraph` functions modify an ETS table behind the scenes so it looks side-effecty like this.

File: lib/shoehorn/release.ex (hunk 1 above)
Comment (fhunleth) [reply lib/shoehorn/release.ex:149]: Not needed. I just was sad to remove it quickly since you can just inspect a digraph to see it.

**Review by jjcarstens — APPROVED** [review]

**Review by ConnorRigby — APPROVED** [review]

**Pull request merged**